### PR TITLE
New integration tests framework

### LINKED ISSUE #3883: Framework for statistical tests
Currently we have frameworks for unit testing, integration testing and sample testing. They are all deterministic, either because the functionality being tested is deterministic, or RNG seeds (both in numpy and espresso) have been set such that the functionality is deterministic. We currently do not have a suitable framework for testing functionality that is stochastic in nature and for which a statistic (e.g. the mean) is known to converge for long simulation times, forcing us to use workarounds, e.g. disable tests in CMake for specific build types, increase test timeout in CI, or reduce the number of integration loops at the expense of accuracy.

**Problem statement**
Several python integration tests are actually not integration tests, but stochastic tests. They do not contribute to code coverage and do not benefit from C++ static assertions and sanitizers, since the underlying functionality already has integration tests. Yet they suffer from significant slowdowns (2-5 times larger runtime) in coverage and sanitizer builds. The Clang sanitizer CI job for example has crossed the 1 hour runtime threshold on multiple occasions over the last few months, prompting us to refactor slow tests to make them run faster at the expense of larger tolerance values in the checks, as well as refactor *unrelated* python tests to make them run faster and gain a few minutes to compensate for slow tests. Statistical tests also tend to undergo large runtime fluctuations on coyote runners that are under load from multiple CI jobs.

Although we should always prefer deterministic tests over statistical tests, there is a need for statistical tests that will not go away. Here are a few candidates for a statistical test framework (times in seconds; `RelWithAssert` builds are compiled with `-O3 -g`):

| Test name | [RelWithAssert](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/250628) | [Coverage](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/250618) | [Sanitizers](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/250626) | Remarks |
| --- | --- | --- | --- | --- |
`langevin_thermostat` | 12.47 | 46.44 | 82.87 | |
`brownian_dynamics` | 6.57 | 14.24 | 23.92 | |
`dpd` | 8.28 | 31.25 | 43.49 | |
`integrator_npt` | 15.65 | 77.96 | 122.07 | |
`lb` | 10.76 | 21.44 | 47.73 | mass & momentum conservation |
`lb_pressure_tensor_acf` | 21.60 | 24.73 | 41.54 | CPU LB is untested (too slow) |
`lb_stokes_sphere` | 3.37 | 39.41 | 99.10 | |
`virtual_sites_tracers` | 5.41 | 49.39 | 111.55 | |
`mass-and-rinertia_per_particle` | 11.45 | 53.08 | 58.30 | |
`rotational-diffusion-aniso` | 18.86 | 97.10 | 81.72 | |
`coulomb_tuning` | 16.97 | 50.51 | 88.34 | P3M tuning is already tested in `p3m_tuning_exceptions` |
`wang_landau_reaction_ensemble` | 115.69 | >300.0 | 532.04 | removed from Coverage builds |

**Prior work**
The sample and tutorial tests were stochastic from January 2019 to July 2020 using randomized numpy seeds and espresso seeds (in 4.1). These tests were too unstable and were eventually converted to deterministic test due to the high failure rate in CI. Generating reliable tolerance intervals was also problematic due to the variety of probability distributions underlying the statistics being tested, forcing us to run new tests in a loop to bootstrap the tolerance interval. This strategy doesn't work for us, so we cannot simply convert statistical tests to pseudo-sample tests to get improved runtimes from `Release` builds.

**Proposed solution**
Re-purpose the unused (and no longer up-to-date) `check_python_skip_long` CMake target to skip the tests mentioned above in coverage and sanitizer builds. When a statistical test provides coverage, move the relevant part into a self-contained integration test. This should remove 11 min (= 22 min * cores) from the Clang job and provide a framework for statistical tests.

**Shortcomings**
One drawback is that integration tests and statistical tests will still run together under the umbrella term of `python` tests. Moving statistical tests to their own CMake target would offer better separation and prevent statistical tests from running if the underlying features raised e.g. ASAN or UBSAN errors, at the expense of more CMake code and a bottleneck in CTest (right now CTest schedules GPU tests to run in serial and CPU tests to take the remaining available cores; introducing a synchronization barrier for statistical tests makes resources allocation less efficient).

### PR DESCRIPTION
**Pull request by jngrad**
Fixes #3883

Description of changes:
- write more thorough RNG checks for thermostats
- skip slow/statistical tests in coverage and sanitizer builds
- fix broken ifdefs for Brownian Dynamics
- cleanup IBM code

### CODE REVIEW

**Comment by jngrad** [comment]
Here is the summary for the python testsuite, using the same CI runner when possible:

| CI job           | Old runtime | New runtime | speed-up  |
| ---------------- |-------------| ----------- | --------- |
| clang-sanitizer  | 2109.2s     | 988.1s      | 19m (53%) |
| no_rotation      | 647.7s      | 468.5s      | 3m (28%)  |
| maxset           | 961.6s      | 616.0s      | 6m (36%)  |
| default          | 970.5s      | 771.3s      | 3m (20%)  |
| cuda11-maxset    | 1113.4s     | 774.9s      | 6m (30%)  |

The complete pipeline runtime went from 72 min to 38 min. The Clang job completes in 35 min instead of 50 min.

@RudolfWeeber your plan in https://github.com/espressomd/espresso/issues/3883#issuecomment-716630623 was really good, however I had to deviate from it slightly:
* `virtual_sites_tracers` is needed to get code coverage for IBM particle coupling to the fluid (not sure why though, the LB actor is removed before the IBM bonds are created)
* Brownian/NpT were moved fully, but new files with minimum checks were introduced for both integrators, otherwise their code coverage would be 0%

Other remarks:
* `make check_python` still works the same way as before
* `make check_python_skip_long` to run the subset of fast tests
* the comment lines in the thermostat tests about RNG seed and counter state were obtained using [cout.diff](https://github.com/espressomd/espresso/files/5568591/cout.diff.txt)

**Comment by jngrad** [comment]
The slowest tests are now:
* `stokesian_dynamics`: 1 min 10 sec in sanitizer, 40 sec in coverage, 20-30s in others
   * now skipped in sanitizer/coverage builds
   * new test added to check the SD interface
* `elc_vs_analytic`: 1 min 20 sec in sanitizer/coverage
   * removing it from coverage builds would require writing a small test without constant potential to preserve code coverage
* reaction ensemble tests:
   * `constant_pH`: 1 min 30 sec in sanitizer, 45 sec in coverage
   * `reaction_ensemble`: 30-40 sec in sanitizer/coverage, 5 sec in others
   * `widom_insertion`: 30-40 sec in sanitizer/coverage, 5 sec in others
   * `wang_landau_reaction_ensemble`: already skipped in sanitizer/coverage
   * long-term plan would be to skip these tests in sanitizer/coverage and write much smaller tests to check the RE interface (error messages, output files, and run a few reaction steps), like I did here for the thermostats and integrators

File: testsuite/python/elc.py [hunk 1]
```diff
@@ -35,8 +35,8 @@ class ElcTest(ut.TestCase):
     def test_finite_potential_drop(self):
         s = self.system
 
-        p1 = s.part.add(pos=GAP + [0, 0, 1], q=+1)
-        p2 = s.part.add(pos=GAP + [0, 0, 9], q=-1)
+        p1 = s.part.add(pos=[0, 0, 1], q=+1)
```
Comment (RudolfWeeber) [inline testsuite/python/elc.py:38]: I thought it doesn't matter, where in the system the gap is? @reinaual 
Did this change influence the test outcome?

**Review by RudolfWeeber — APPROVED** [review]

File: testsuite/python/elc.py (hunk 1 above)
Comment (jngrad) [reply testsuite/python/elc.py:38]: This change did not alter the numerical result. I think the gap is always at the higher value of the z-axis:
https://github.com/espressomd/espresso/blob/127ad4ef7c9cee2b4242e76139cf69f55230d628/src/core/electrostatics_magnetostatics/elc.hpp#L92-L93
https://github.com/espressomd/espresso/blob/127ad4ef7c9cee2b4242e76139cf69f55230d628/src/core/electrostatics_magnetostatics/elc.cpp#L1196

If particles enter the gap, they are misclassified as particles from the dielectric contrast when calculating the forces and energies. The documentation warns against using ELC without walls to block particles from entering the gap:
https://github.com/espressomd/espresso/blob/127ad4ef7c9cee2b4242e76139cf69f55230d628/doc/sphinx/electrostatics.rst#L254-L259

Comment (reinaual) [reply testsuite/python/elc.py:38]: In theory it should not matter, to my knowledge in the implementation it does. The lower interface is assumed at `z=0` and the upper one at `z=h`, which is relevant for the error-estimation and the placement of the image charges.
To my understanding the testcase should be correct for every given prefactor, however setting it to `1e-100` allows to explicitly test the potential difference contribution but hides the misplaced particles in this case.

**Pull request merged**